scripts/migration/lib.sh

```diff
@@ -0,0 +1,286 @@
+
+log_info_no_newline() {
+  echo -e -n "[\e[34mck8s\e[0m] ${CK8S_STACK}: ${*}" 1>&2
+}
+
+log_info() {
+  log_info_no_newline "${*}\n"
+}
+
+log_warn_no_newline() {
+  echo -e -n "[\e[33mck8s\e[0m] ${CK8S_STACK}: ${*}" 1>&2
+}
+
+log_warn() {
+  log_warn_no_newline "${*}\n"
+}
+
+log_error_no_newline() {
+  echo -e -n "[\e[31mck8s\e[0m] ${CK8S_STACK}: ${*}" 1>&2
+}
+
+log_error() {
+  log_error_no_newline "${*}\n"
+}
+
+log_fatal() {
+  log_error "${*}"
+  exit 1
+}
+
+# --- git version
+
+git_version() {
+  git -C "${ROOT}" describe --exact-match --tags 2> /dev/null || git -C "${ROOT}" rev-parse HEAD
+}
+
+# --- config functions ---
+
+# Usage: config_version <sc|wc>
+config_version() {
+  if [[ ! "${1:-}" =~ ^(sc|wc)$ ]]; then
+    log_fatal "usage: config_version <sc|wc>"
+  fi
+
+  local prefix="${1}"
+
+  local version
+  version="$(yq4 ".global.ck8sVersion" <<< "${CONFIG["${prefix}"]}")"
+
+  VERSION["${prefix}-config"]="${version}"
+  version="${version#v}"
+  VERSION["${prefix}-config-major"]="${version%%.*}"
+  version="${version#*.}"
+  VERSION["${prefix}-config-minor"]="${version%%.*}"
+  version="${version#*.}"
+  VERSION["${prefix}-config-patch"]="${version}"
+}
+
+# Usage: config_validation <secrets|sc|wc>
+config_validate() {
+  local defaults
+  local setmes
+  local pass="true"
+
+  IFS=$'\n'
+
+  case "${1:-}" in
+  secrets)
+    log_info "validating ${1}"
+
+    setmes="$(yq_paths "set-me" < <(sops -d "${CK8S_CONFIG_PATH}/secrets.yaml"))"$'\n'"$(yq_paths "somelongsecret" < <(sops -d "${CK8S_CONFIG_PATH}/secrets.yaml"))"
+    for setme in ${setmes}; do
+      log_error "error: \"${setme//\"/}\" is unset in ${1}"
+      pass="false"
+    done
+    ;;
+
+  sc|wc)
+    log_info "validating ${1}-config"
+
+    defaults="$(yq_merge "${CK8S_CONFIG_PATH}/defaults/common-config.yaml" "${CK8S_CONFIG_PATH}/defaults/${1}-config.yaml")"
+    setmes="$(yq_paths "set-me" <<< "${defaults}")"
+
+    for setme in ${setmes}; do
+      compare=$(diff <(yq4 -oj "${setme}" <<< "${defaults}") <(yq4 -oj "${setme}" <<< "${CONFIG["${1}"]}") || true)
+      if [[ -z "${compare}" ]]; then
+          log_error "error: \"${setme//\"/}\" is unset in ${1}-config"
+          pass="false"
+      fi
+    done
+    ;;
+
+  *)
+    log_fatal "usage: config_validate <secrets|sc|wc>"
+    ;;
+  esac
+
+  if [[ "${pass}" = "false" ]]; then
+    exit 1
+  fi
+
+  IFS="${GLOBAL_IFS}"
+}
+
+# Usage: config_load <sc|wc>
+config_load() {
+  case "${1:-}" in
+  sc)
+    log_info "loading ${1}-config"
+    CONFIG[sc]="$(yq_merge "${SC_CONFIG_FILES[@]/#/"$CK8S_CONFIG_PATH/"}")"
+    config_version sc
+    ;;
+  wc)
+    log_info "loading ${1}-config"
+    CONFIG[wc]="$(yq_merge "${WC_CONFIG_FILES[@]/#/"$CK8S_CONFIG_PATH/"}")"
+    config_version wc
+    ;;
+  *)
+    log_fatal "usage: config_load <sc|wc>"
+    ;;
+  esac
+}
+
+check_sops() {
+  grep -qs "sops:\\|\"sops\":\\|\\[sops\\]\\|sops_version=" "${1:-/dev/null}"
+}
+
+check_config() {
+  if [ -z "${THIS:-}" ]; then
```

Comment:
Do we really need this since we have `set -u`? Then it should error out when expanding those

Reply:
No, but it looks nicer when things do go wrong.